lib/streamlit/connections/sql_connection.py

```diff
@@ -27,33 +29,44 @@
     from sqlalchemy.orm import Session
 
 
+_ALL_CONNECTION_PARAMS = {
+    "url",
+    "driver",
+    "dialect",
+    "username",
+    "password",
+    "host",
+    "port",
+    "database",
+}
 _REQUIRED_CONNECTION_PARAMS = {"dialect", "username", "host"}
 
 
 class SQL(ExperimentalBaseConnection["Engine"]):
     def _connect(self, autocommit: bool = False, **kwargs) -> "Engine":
         import sqlalchemy
 
-        secrets = self._secrets
+        kwargs = extract_from_dict(_ALL_CONNECTION_PARAMS, deepcopy(kwargs))
+        conn_params = merge_dicts([self._secrets.to_dict(), kwargs])
 
-        if "url" in secrets:
-            url = sqlalchemy.engine.make_url(secrets["url"])
+        if "url" in conn_params:
+            url = sqlalchemy.engine.make_url(conn_params["url"])
         else:
             for p in _REQUIRED_CONNECTION_PARAMS:
-                if p not in secrets:
+                if p not in conn_params:
                     raise StreamlitAPIException(f"Missing SQL DB connection param: {p}")
 
-            drivername = secrets["dialect"] + (
-                f"+{secrets['driver']}" if "driver" in secrets else ""
+            drivername = conn_params["dialect"] + (
+                f"+{conn_params['driver']}" if "driver" in conn_params else ""
             )
 
             url = sqlalchemy.engine.URL.create(
                 drivername=drivername,
-                username=secrets["username"],
-                password=secrets.get("password"),
-                host=secrets["host"],
-                port=int(secrets["port"]) if "port" in secrets else None,
-                database=secrets.get("database"),
+                username=conn_params["username"],
+                password=conn_params.get("password"),
+                host=conn_params["host"],
+                port=int(conn_params["port"]) if "port" in conn_params else None,
+                database=conn_params.get("database"),
             )
 
         eng = sqlalchemy.create_engine(url, **kwargs)
```

Comment:
snowflake-sqlalchemy uses a `connect_args` kwarg outside the URL to specify common behaviors like non-password based authentication ([rough cite](https://docs.snowflake.com/developer-guide/python-connector/sqlalchemy#key-pair-authentication-support) - this is also what I found when I tried it out).

Similarly, bigquery-sqlalchemy uses a `credentials_info` kwarg which seems to be required for any usage ([cite](https://github.com/googleapis/python-bigquery-sqlalchemy#authentication)).

It seems like there's no way to get that passed into create_engine directly from secrets with this implementation. So to use these dialects (which are pretty important!) I still have to do something like:
`conn = st.connection('sql', connect_args=dict(st.secrets.snowflakeargs))`

I would _love_ to be able to just specify a secrets section like `[connections.sql.connect_args]` and have it get passed through.

Reply:
(Should I move this comment to the main feature PR? Since I guess it doesn't need to be addressed in this PR)

Reply:
Ack -- will see if I can address this early next week